src/sorting.jl

```diff
@@ -0,0 +1,392 @@
+"""
+Quicksort!
+Alex Ellison
+@xaellison
+
+Usage:
+quicksort!(my_cuarray)
+
+The main quicksort kernel uses dynamic parallelism. Let's call blocksize M. The
+first part of the kernel bubble sorts M elements with maximal stride between
+lo and hi. If the sublist is <= M elements, stride = 1 and no recursion
+happens. Otherwise, we pick element lo + M / 2 * stride as a pivot. This is
+an efficient choice for random lists and pre-sorted lists.
+
+Partition is done in stages:
+1. In a separate kernel: Merge-sort batches of M values using their comparison
+    to pivot as a key. The comparison alternates between < and <= with recursion
+    depth. This makes no difference when there are many unique values, but when
+    there are many duplicates, this effectively partitions into <, =, and >.
+2. Consolidate batches. This runs inside the quicksort kernel.
+
+Information about the partitions of each batch is stored in shared memory if
+there are fewer than N batches (list length <= M ^ 2). Otherwise, it is stored in
+a global array. Any sub-kernels which also need that array will be executed
+synchronously so they don't conflict over that memory.
+
+Naming conventions:
+Sublists (ranges of the list being sorted) are denoted by `my_floor` and one of
+    `L` and `my_ceil`. `my_floor` is an exclusive lower bound, `my_ceil` is an
+    inclusive upperboard, `L` is their difference.
+
+`lo` / `hi` are used in place of `my_floor` / `my_ceil` in `qsort_kernel`
+    `lo` / `hi` are indices of  batches in `consolidate_batch_partition`
+
+`b_sums` is "batch sums", the number of values in a batch which are >= pivot
+"""
+
+#-------------------------------------------------------------------------------
+# Integer arithmetic
+"""
+Equivalent to Int(ceil(a / b))
+"""
+function ceil_div(a, b)
+    return a ÷ b + 1 & (a % b != 0)
+end
+
+"""
+Returns smallest power of 2 >= x
+"""
+function pow2_ceil(x)
+    out = 1
+    while out < x
+        out *= 2
+    end
+    out
+end
+
+function pow2_floor(x)
+    out = 1
+    while out * 2 <= x
+        out *= 2
+    end
+    out
+end
```

Comment:
i don't know about numerical math like this but I assume this is faster than `floor(log(2, x))`
Seems to be not much difference in speed

```
function pow2_ceil(x)
    out = 1
    while out < x
        out *= 2
    end
    out
end

function pow2_ceil_shift(x)
    out = 1
    while out < x
        out <<= 1
    end
    out
end

using BenchmarkTools

function pow2_ceil_v2(x)
    2^ceil(Int, log(2, x))
end


x = abs(Float32(rand(Int16)))

pow2_ceil(x) == pow2_ceil_shift(x)
pow2_ceil(x) == pow2_ceil_v2(x)

@benchmark pow2_ceil($x)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     25.074 ns (0.00% GC)
  median time:      25.276 ns (0.00% GC)
  mean time:        25.721 ns (0.00% GC)
  maximum time:     81.946 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     997

@benchmark pow2_ceil_shift($x)

BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     25.050 ns (0.00% GC)
  median time:      25.251 ns (0.00% GC)
  mean time:        25.636 ns (0.00% GC)
  maximum time:     81.864 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     998


@benchmark pow2_ceil_v2($x)

BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     19.038 ns (0.00% GC)
  median time:      20.542 ns (0.00% GC)
  mean time:        21.411 ns (0.00% GC)
  maximum time:     200.600 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     998
```

Reply:
`ceil_div` can be replaced by `cld`.

You can use this code as replacements to avoid iterative algorithms.
```julia
@inline function pow2_floor_gpu(x)
    CUDA.popc(x) != 1 && return 1 << (sizeof(x)*8 - CUDA.ffs(CUDA.brev(x)))
    return x
end

@inline function pow2_ceil_gpu(x)
    CUDA.popc(x) != 1 && return 1 << (sizeof(x)*8 - CUDA.ffs(CUDA.brev(x)) + 1)
    return x
end
```